test/functional/testlib.bash

```diff
@@ -0,0 +1,502 @@
+# If no existing file is specified to point to, a new file will be created and pointed to
+# by the link.
+# Parameters:
+# - PATH: the path where the symbolic link will be created
+# - FILE: the path to the file to point to 
+create_link() { 
+
+	local path=$1
+	local pfile=$2
+	local hashed_name
+	validate_path "$path"
+	
+	# if no file is specified, create one
+	if [ -z "$pfile" ]; then
+		pfile=$(create_file "$path")
+	else
+		validate_item "$pfile"
+	fi
+	sudo ln -rs "$pfile" "$path"/testlink
+	hashed_name=$(sudo "$SWUPD" hashdump "$path"/testlink 2> /dev/null)
+	sudo mv "$path"/testlink "$path"/"$hashed_name"
+	echo "$path/$hashed_name"
+
+}
+
+# Creates a tar for the specified item in the same location
+# Parameters:
+# - ITEM: the relative path to the item (file, directory, link, manifest)
+create_tar() {
+
+	local item=$1
+	local path
+	local item_name
+	validate_item "$item"
+
+	path=$(dirname "$(realpath "$item")")
+	item_name=$(basename "$item")
+	# if the item is a directory exclude its content when taring
+	if [ -d "$item" ]; then
+		sudo tar -C "$path" -cf "$path"/"$item_name".tar --exclude="$item_name"/* "$item_name"
+	else
+		sudo tar -C "$path" -cf "$path"/"$item_name".tar "$item_name"
+	fi
+
+}
+
+# Creates an empty manifest in the specified path
+# Parameters:
+# - PATH: the path where the manifest will be created
+# - BUNDLE_NAME: the name of the bundle which this manifest will be for
+create_manifest() {
+
+	local path=$1
+	local name=$2
+	local version
+	validate_path "$path"
+	validate_param "$name"
+
+	version=$(basename "$path")	
+	echo -e "MANIFEST\\t3" | sudo tee "$path"/Manifest."$name" > /dev/null
+	echo -e "version:\\t$version" | sudo tee -a "$path"/Manifest."$name" > /dev/null
+	echo -e "previous:\\t0" | sudo tee -a "$path"/Manifest."$name" > /dev/null
+	echo -e "filecount:\\t0" | sudo tee -a "$path"/Manifest."$name" > /dev/null
+	echo -e "timestamp:\\t$(date +"%s")" | sudo tee -a "$path"/Manifest."$name" > /dev/null
+	echo -e "contentsize:\\t0" | sudo tee -a "$path"/Manifest."$name" > /dev/null
+	echo "" | sudo tee -a "$path"/Manifest."$name" > /dev/null
+	echo "$path/Manifest.$name"
+	
+}
+
+# Adds the specified item to an existing bundle manifest
+# Parameters:
+# - MANIFEST: the relative path to the manifest file
+# - ITEM: the relative path to the item (file, directory, symlink) to be added
+# - ITEM_PATH: the absolute path of the item in the target system when installed
+add_to_manifest() { 
+
+	local manifest=$1
+	local item=$2
+	local item_path=$3
+	local item_type
+	local item_size
+	local name
+	local version
+	local filecount
+	local contentsize
+	local linked_file
+	validate_item "$manifest"
+	validate_item "$item"
+	validate_param "$item_path"
+	
+	item_size=$(du -b "$item" | cut -f1)
+	name=$(basename "$item")
+	version=$(basename "$(dirname "$manifest")")
+	# add to filecount
+	filecount=$(sudo cat "$manifest" | grep filecount | awk '{ print $2 }')
```

Comment:
Any reason to suspect that $manifest is not readable? If not then
`filecount=$(awk '/filecount/ { print $2}' "$manifest")`

If it might not be readable then either add a `sudo` or a `sudo cat | awk` as taste dictates.

Reply:
done